### ISSUE TITLE
Querying security groups slow (serial) when having many of them

### ISSUE THREAD

**Issue by vlerenc** [cloudfoundry/cli#892]
Hi,

when running cf security-groups on a system with many security-groups (>2K) the response time of the CF CLI gets slow (1-2 minutes for 2K security groups).

@XenoPhex said in response to #878:

> At the moment, we make a list request for security groups (in series) and then do a spaces list request for each security group (in series). This means if you have 2006 security groups, we're doing at least 2006/50 + 2006 requests to generate that table.
> Right now we have no plans on making these requests in parallel, which would really speed things up. But if this is a significant issue for you, please feel free to file a separate ticket and we can look into possible solutions!

Given that we make heavy use of security groups for service instance isolation it is kind of an issue for the ops colleagues managing our larger landscapes.

Cheers, Vedran
### Command

cf security-groups
### CLI Version

cf version 6.20.0+198f496-2016-07-12
### Error

Performance slow due to serial access.
### CC API Endpoint Version

API endpoint: https://api.cf.service-fabrik.sc6.sapcloud.io (API version: 2.55.0)
### CF Trace

Very large, many consequent calls.
### Platform Details

Distributor ID: Ubuntu
Description:    Ubuntu 14.04.4 LTS
Release:        14.04
Codename:       trusty
### Shell

Terminal/Bash

**Comment by dkoper** [comment]
Thanks for submitting the report.
@SocalNick Do you have any suggestions? A different API?

**Comment by SocalNick** [comment]
@vlerenc you are creating a security group for each service instance? I don't think that's what we had in mind for security groups / service instances. this might be an interesting conversation to take to cf-dev mailing list to see if other operators have any ideas how to accomplish this.

curious what you list all 2000 security groups for? do you need to audit what orgs / spaces are assigned to them?

@dkoper does `cf orgs` have a max before it doesn't list them all? could we consider leaving off the organization / space when there are too many security groups? are paging flags an anti-pattern in the CLI?

**Comment by vlerenc** [comment]
@SocalNick Yes, we create a sec grp per service instance. Why? Because we have automated our service instance provisioning on Docker, Bosh, etc. When our customers create service instances, we provision them automatically and part of that is also creating and deleting sec grps for isolation.

@SocalNick No, it's not really a use case to list all of them, but ops must have access to the sec grps. So they should at least be able to list them timely and apply unix tools to filter for what they need to look at for instance.

@SocalNick 2K was just a number. One of our landscapes had some 2K+ service instances and when we implemented isolation based on sec grps we wanted to check whether it would work with the current landscape. It didn't at first because of the crash and now it does, but it is kind of slow.

**Comment by SocalNick** [comment]
cc @benlaplanche @dieucao @avade RE: security group per service instance

**Comment by dkoper** [comment]
@SocalNick `cf orgs` has no max. `cf login` is the only command I'm aware of that stops at 50. All other commands are intended to be non-interactive, so changing this now would be a breaking change.
We could consider a way to display how many elements the CLI is going to display, so users can ctrl-c if they feel it may take to look, but I suppose they can ctrl-c already, and operators probably know their deployment well enough to know (at least they will after one try) that they have a lot of entries (so would avoid running the command if they could help it).
If the command is slow because we display too much information (i.e. if we could reduce the number of API calls made by omitting non-essential data), we could look into that.

**Comment by SocalNick** [comment]
@dkoper as @XenoPhex mentioned, the calls needed to retrieve the data for `X Security Groups` is  `X/50 + X`. If there aren't upper limits on `X`, I think we should consider leaving off the additional data  necessitating the the `+ X` part of that equation.

**Comment by dkoper** [comment]
It seems to speed this up we need to omit information that we are currently displaying.
The question is what information that is, and whether leaving that out by default can lead to compatibility issues.
Other high priority stories have pushed this investigation on the shelf.

**Comment by vlerenc** [comment]
Sorry to hear, it really is a pain point on our side.

**Comment by sklevenz** [comment]
Unfortunately this is an issue in production and the api call just fails. 

Ideally such mass calls never fail due to a restricted paging supported by the server or other mechanism. A simple solution could be the "cf sercurity-groups" returns only few data like sg-name or sg-id. 
A client can take name or id and decide to query for additional information in a second call only if required.

WDYT?

**Comment by dkoper** [comment]
@sklevenz What do you mean the API call just fails? As noted, the CLI makes many API calls, but each of them individually should not fail.

The issue I see with changing this command's output now, for a command that has been there for ages, is that there may be plenty of users (and scripts!) for whom its working fine and they depend on (e.g. parse) the current output and would then break.

Which brings us back to the use case: you want to first list the existing security groups by name (because you forgot their names but will recognize the one you are after when you see it? Could a filter help you here?) and then potentially query the details of only that security group?

**Comment by dkoper** [comment]
@vlerenc We have just finished refactoring this command. The purpose of the refactor was not to improve its speed, but as the API calls have changed I would expect a difference in performance.

Maybe you could do a before/after comparison for your environment to help us understand where we are now?
For before, you can use any current cf CLI release.
For after, use [the current edge binary](https://github.com/cloudfoundry/cli#edge-binaries) with environment variable `CF_CLI_EXPERIMENTAL=true`.  You can be sure you're running the refactored command if the first line ends with `...`.

Cheers,
Dies Koper
CF CLI PM

**Comment by dkoper** [comment]
Closing, pending a new performance evaluation.

**Comment by sklevenz** [comment]
@abbyachau I don't see a change in behavior. The command fails after 10 minutes:

```
time cf security-groups
Getting security groups as admin_cf...
Error unmarshalling the following into a cloud controller error: 404 Not Found: Requested route ('api.cf.sap.hana.ondemand.com') does not exist.

FAILED

real	10m27.911s
user	0m5.640s
sys	0m5.940s
```
The server simply returns to much data.

This is a general problem of any REST api. Maybe you can have a look at OData (https://www.odata.org/) which is a standard for REST API addressing such issues.
